plenum/server/consensus/consensus_data_provider.py

```diff
@@ -75,70 +75,72 @@ def stable_checkpoint(self) -> int:
+        if pp in self._preprepared:
+            raise LogicError('New pp cannot be stored in preprepared')
+        if pp in self._prepared:
+            raise LogicError('New pp cannot be stored in prepared')
+        if self._checkpoints and pp.ppSeqNo < self._checkpoints[-1].seqNoEnd:
+            raise LogicError('ppSeqNo cannot be lower than last checkpoint')
+
         self._preprepared.append(pp)
 
     def prepare_batch(self, pp: PrePrepare):
         """
         After prepared certificate for pp had collected,
         it removed from _preprepared and placed into _prepared list
         """
-        if pp is None:
-            raise LogicError('pp does not exist')
+        # TODO: move logic checks upper the functions stack
         if pp not in self._preprepared:
             raise LogicError('Unprepared pp must be stored in preprepared')
         self._preprepared.remove(pp)
         self._prepared.append(pp)
 
-    def free_batch(self, pp):
+    def clear_batch(self, pp: PrePrepare):
         """
-        When 3pc batch had committed, rejected or catchuped till, it removed from _prepared list
+        When 3pc batch processed, it removed from _prepared list
         """
-        if pp is None:
-            raise LogicError('pp does not exist')
-        if pp in self._preprepared:
-            raise LogicError('Prepared pp cannot be stored in preprepared')
-        if pp not in self._prepared:
-            raise LogicError('Prepared pp must be stored in prepared')
+        # TODO: move logic checks upper the functions stack
+        if pp in self._preprepared == pp not in self._prepared:
+            raise LogicError('Batch can be only in one condition while clearing ')
         self._prepared.remove(pp)
 
-    def free_all(self):
+    def clear_all_batches(self):
         """
-        When catchup finished, backups clear up their pre-prepares
+        Clear up all preprepared and prepared
         """
         self._prepared.clear()
         self._preprepared.clear()
 
-    def get_3pc(self):
-        return self.view_no, self.pp_seq_no
-
-    def set_3pc(self, key):
-        if len(key) != 2:
-            raise LogicError('3pc key must be a pair')
-        self.view_no = key[0]
-        self.pp_seq_no = key[1]
-
     def append_checkpoint(self, checkpoint: Checkpoint):
-        self._checkpoints[(checkpoint.seqNoStart, checkpoint.seqNoEnd)] = checkpoint
+        self._checkpoints.add(checkpoint)
 
-    def remove_checkpoint(self, key):
-        if key not in self._checkpoints:
-            raise LogicError('key for removal must be in checkpoints')
-        del self._checkpoints[key]
+    def remove_checkpoint(self, end_seq_no):
+        new_checkpoints = [c for c in self._checkpoints if c.seqNoEnd != end_seq_no]
+        if len(new_checkpoints) != len(self._checkpoints) - 1:
+            raise LogicError('One checkpoint needed to be removed')
+        self._checkpoints = new_checkpoints
```

Comment:
Do we actually need to make this public function? As far as I understand we should remove checkpoints automatically when we increase stable checkpoint.